linters/importorder/main.go

```diff
@@ -0,0 +1,465 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/build"
+	"go/parser"
+	"go/token"
+	"log"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/kisielk/gotool"
+	"golang.org/x/tools/go/loader"
+)
+
+const (
+	standardImportGroup = "STDLIB"
+	externalImportGroup = "EXTERNAL"
+)
+
+var (
+	errMultipleImport       = errors.New("more than one import declaration found")
+	errOutOfOrder           = errors.New("import is out of order")
+	errImportMatchedAlready = errors.New("import already matched pattern")
+	errGroupMachedAlready   = errors.New("import group already matches previously seen pattern")
+	errNoMatch              = errors.New("import does not match any of the provided patterns or number of groups exceeds number of patterns provided")
+
+	defaultPattern = fmt.Sprintf("%s %s", standardImportGroup, externalImportGroup)
+)
+
+type lintError struct {
+	fileName    string
+	importName  string
+	line        int
+	patternSeen string
+	err         error
+}
+
+type lintErrors []lintError
+
+func main() {
+	tags := flag.String("tags", "", "List of build tags to take into account when linting.")
+	skipVendor := flag.Bool("skip-vendor", true, "Skip vendor directors.")
+	rawPatterns := flag.String("patterns", defaultPattern, "Specify the patterns of each group in order. If checking for Go standard imports write `STDLIB`, if checking for a wildard group write `EXTERNAL`.")
+
+	flag.Parse()
+	importPaths := gotool.ImportPaths(flag.Args())
+	if len(importPaths) == 0 {
+		flag.Usage()
+		return
+	}
+
+	patterns := strings.Fields(*rawPatterns)
+	if len(patterns) < 1 {
+		log.Fatal("List of patterns must be greater than 0\n")
+	}
+
+	filteredPaths := importPaths
+	if *skipVendor {
+		filteredPaths = filterOutVendor(filteredPaths)
+	}
+
+	groupedErrors := handleImportPaths(filteredPaths, strings.Fields(*tags), patterns)
+	printErrors(groupedErrors)
+}
+
+func printErrors(groupedErrors lintErrors) {
+	if len(groupedErrors) != 0 {
```

Comment:
don't need the `if !=` check here. range over an empty slice works as you'd expect - e.g. https://play.golang.org/p/9CyPK4p0QKj